### PR TITLE
Rotation undo transform for Aleatoric uncertainty

### LINKED ISSUE #136: undo_transform of Affine and Elastic transformations
To use the aleatoric uncertainty, we need to implement some new "undo_transform", such as for the `Affine` and the `Elastic` transformations.

### PR DESCRIPTION
**Pull request by charleygros**
We need to implement Rotation undo transform for Aleatoric uncertainty purposes. See issue #136 

Sister PR: https://github.com/neuropoly/medicaltorch/pull/16

To test this PR:
1. Run:
```
python testing/test_undoTrans.py
```
2. Use `cmd_test` with:
- `uncertainty: {aleatoric: true}`.
- in `transformation_training`: `RandomRotation: {degrees: XX}`

Fix #136

### CODE REVIEW

**Comment by charleygros** [comment]
Important note: I have introduced `transformation_testing` in the config file. Reason: when using aleatoric MC, the DA differs from the validation (no DA) and from the training (Elastic or RandomAffine have no undo_transform, which is needed during inference).

**Comment by charleygros** [comment]
> Important note: I have introduced `transformation_testing` in the config file. Reason: when using aleatoric MC, the DA differs from the validation (no DA) and from the training (Elastic or RandomAffine have no undo_transform, which is needed during inference).

To make it explicit, the transformations are now displayed in the terminal:
```
Applied transformations:
	- Resample
	- ROICrop2D
	- ToTensor
	- NormalizeInstance
```

File: ivadomed/main.py [hunk 1]
```diff
@@ -852,8 +858,13 @@ def cmd_test(context):
 
 
 def cmd_eval(context):
+    path_pred = os.path.join(context['log_directory'], 'pred_masks')
+    if not os.path.isdir(path_pred):
+        print('\nRun Inference\n')
+        cmd_test(context)
+    print('\nRun Evaluation on {}\n'.format(path_pred))
```
Comment (charleygros) [inline ivadomed/main.py]: If the inference has not been run before running the evaluation --> then we call `cmd_test`.

Which means that we only have to run `cmd_eval` if `pred_masks` folder does not exist: "kill two birds with one stone" JCA.

**Comment by charleygros** [comment]
I moved all the "uncertainty" code from `cmd_test` to `utils` --> we are just calling `run_uncertainty` in `cmd_test`.

**Comment by charleygros** [comment]
Preliminary results: on the testing set, median and IQR.

(Note: Ideal score for Dice and LTPR are 100., while we aim at minimizing LFDR)

No Epistemic & No Aleatoric:
```
3D Dice: 52.4 (0 - 58.2) %
3D LFDR: 33.3 (25 - 50) %
3D LTPR: 87.5 (68.8 - 100) %
```

Epistemic & No Aleatoric:
```
3D Dice: 38.4 (0 - 58.3) %
3D LFDR: 63.6 (39.3 - 78.9) %
3D LTPR: 75 (68.7 - 93.8) %
```

No Epistemic & Aleatoric (RandomRotation, 20°):
```
3D Dice: 47.2 (0 - 55.3) %
3D LFDR: 38.8 (34.4 - 60) %
3D LTPR: 87.5 (43.8 - 100) %
```

Preliminary experiments conclusion: We are gaining information with uncertainty / soft prediction but the first results are showing less good results on hard masks compared to "no Dropout & DA at testing time".

Will keep digging & keep you posted.

**Comment by charleygros** [comment]
New results on the same dataset / conditions:

Preliminary results: on the testing set, median and IQR.

(Note: Ideal score for Dice and LTPR are 100., while we aim at minimizing LFDR)

No Epistemic & No Aleatoric:
```
3D Dice: 52.4 (0 - 58.2) %
3D LFDR: 33.3 (25 - 50) %
3D LTPR: 87.5 (68.8 - 100) %
```

Epistemic & No Aleatoric:
```
3D Dice: 38.4 (0 - 46.1) %
3D LFDR: 72.4 (64.7 - 74.6) %
3D LTPR: 87.5 (68.8 - 100) %
```

No Epistemic & Aleatoric (RandomRotation, 30°):
```
3D Dice: 43.8 (0 - 53.4) %
3D LFDR: 45 (40 - 66.1) %
3D LTPR: 100 (81 - 100) %
```

Conclusion: Improved results in terms of lesion wise metrics. But still less than without uncertainty. Keep looking

**Comment by charleygros** [comment]
I run an aleatoric experiment without `RandomRotation` --> it gives the same results than without MC --> expected behaviour, okay, next.

**Comment by charleygros** [comment]
> I run an aleatoric experiment without `RandomRotation` --> it gives the same results than without MC --> expected behaviour, okay, next.

Same result and conclusion with `RandomRotation(degrees=0)`.

**Comment by charleygros** [comment]
When removing the binarisation in `save_nii`:

No Epistemic & No Aleatoric:
```
3D Dice: 52.4 (0 - 58.2) %
3D LFDR: 33.3 (25 - 50) %
3D LTPR: 87.5 (68.8 - 100) %
```

Epistemic & No Aleatoric:
```
3D Dice: 37.2 (0 - 51.3) %
3D LFDR: 63.3 (52.5 - 75) %
3D LTPR: 100 (81.25 - 100) %
```

No Epistemic & Aleatoric (RandomRotation, 30°):
```
3D Dice: 47.7 (0 - 55.5) %
3D LFDR: 39.3 (25.9 - 57.5) %
3D LTPR: 100 (81 - 100) %
```

Conclusion:
- Improved results compared to previous experiment for both aleatoric & epistemic. Give him a chocolate.
- Aleatoric: We now have similar results between with and without TestTimeDA. Expected behaviour --> I feel personally happy with that. Of course, will need to be further validated & dissected.
- Epistemic: why do we have this drop of results? No idea at this point. Should I reduce the `p_dropout`? Am I doing something wrong [here](https://github.com/neuropoly/ivado-medical-imaging/blob/master/ivadomed/main.py#L731)? Will look at it tomorrow. Any opinion more than welcome.

**Comment by charleygros** [comment]
Final results of this PR: on the entire testing set, median and IQR.

(Note: Ideal score for Dice and LTPR are 100., while we aim at minimizing LFDR. Best scores indicated in bold.)

Baseline:
```
3D Dice: **42.8** (**15** - **57.6**) %
3D LFDR: **33.3** (**0** - **60**) %
3D LTPR: 100 (66.7 - 100) %
```

Epistemic (`n_it=20, p_dropout=0.3`):
```
3D Dice: 34.3 (11.4 - 51.7) %
3D LFDR: 66.7 (50 - 81.8) %
3D LTPR: **100** (**88.9** - **100**) %
```

Aleatoric (`n_it=20, RandomRotation, 30°`):
```
3D Dice: 39.3 (13.7 - 55.5) %
3D LFDR: 55.1 (28.6 - 70.4) %
3D LTPR: 100 (80 - 100) %
```

Conclusion: The MC samples introduce some FP which pulls down the overall scores. Future work will focus on implementing post-processing tools benefiting from the uncertainty metrics to mitigate this limitation.

File: config/config_sctTesting_lesionAx.json [hunk 2]
```diff
@@ -73,5 +73,19 @@
...
+                        },
+                "ROICrop2D": {
+                        "size": [48, 48]
+                        },
+                "RandomRotation": {
+                        "degrees": 30
+                        },
+                "ToTensor": {},
+                "NormalizeInstance": {}
+                },
```
Comment (AnBucquet) [inline config/config_sctTesting_lesionAx.json]: I guess we need to add this to all config files tho?

Comment (charleygros) [reply config/config_sctTesting_lesionAx.json]: Done (:

Comment (AnBucquet) [reply config/config_sctTesting_lesionAx.json]: You rock!

**Pull request merged**